contributors/design-proposals/storage/csi-migration.md

```diff
@@ -0,0 +1,443 @@
+
+We will also have one feature flag for each driver’s migration so that each
+driver migration can be turned on and off individually. 
+
+The new feature gates for alpha are:
+```
+// Enables the in-tree storage to CSI Plugin migration feature.
+CSIMigration utilfeature.Feature = "CSIMigration"
+
+// Enables the GCE PD in-tree driver to GCE CSI Driver migration feature.
+CSIMigrationGCE utilfeature.Feature = "CSIMigrationGCE"
+
+// Enables the AWS in-tree driver to AWS CSI Driver migration feature.
+CSIMigrationAWS utilfeature.Feature = "CSIMigrationAWS"
+```
+
+## Translation Layer
+The main mechanism we will use to migrate plugins is translating In-tree plugin
+sources into CSI Sources and redirecting operation calls to the CSI Driver
+instead of the in-tree driver.
+
+Pros:
+* Keeps old API objects as they are
+* Facilitates gradual roll-over to CSI
+
+Cons:
+* Somewhat complicated and error prone.
+* AttachDetach Controller must have permissions to modify PVs.
+* Bespoke translation logic for each in-tree plugin
+
+### Dynamically Provisioned Volumes
+
+#### Kubernetes Changes
+In-tree provisioner will be changed to have a feature gate that when turned on
+will not register the in-tree volume plugin into PV controller. We will modify
+`provisionClaimOperation` to set the claim provisioner to the migrated external
+provisioner name, so the external provisioner will pick this up and treat it as
+a CSI Volume from then on. The controller will automatically annotate PVCs for
+external provisioning by an provisioner that handles e.g. “google.io/gce-pd”.
+
+#### External Component Changes
+External provisioner will use a different  provisioner name as the internal
+volume plugin name (e.g. “google.com/gce-pd”). Note that the CSI driver name can
+be different and this should be just matter of configuration.
+
+Caveat: On downgrade the CSI Driver must still exist to continue the lifecycle
+of the PV’s and volumes it provisioned, the in-tree driver will not pick up
+where the CSI Driver left off. We could have the in-tree volume pick up where
+the CSI driver left off, but in that case the two provisioners race with each
+other if the CSI Driver is still on - this is because there is no consistent way
+(works when CSI driver is both on or off) of communicating between the CSI and
+Internal driver who is in charge of the PV if both contend for ownership of the
+PV.
+
+The CSI Driver replacing the in-tree plugin must support the same storage class
+parameters as the internal volume plugins, so that we continue to have user
+facing consistency. The details of this will not be covered in this design, but
+it will a requirement for admission for migration.
+
+The provisioner will pick up the PVC, and create a PV with a CSI Source that it
+would do normally. Using CSISource in the PV might not be fully “backwards
+compatible” but it will be ok from a user facing behavior change perspective
```

Comment:
This would be an API break.  Tools that rely on the PV source would no longer be able to find that data.  At best, you would have to mirror the data in the api into the CSI source and keep them in sync.